### ISSUE TITLE
Handle OSGi R6 Breaking changes in a backward compatible manner

### ISSUE THREAD

**Issue by davidjgonzalez** [Adobe-Consulting-Services/acs-aem-commons#1631]
We should investigate if/how we can handle the legacy "-" params for OSGi services after the (currently) breaking changes to OSGi R6.

The  most feasible approach (thanks CZ) looks to be:

> There is no way to use the annotation based approach of defining properties if the property name contains a dash. That's a limitation to avoid complicated mapping rules of allowed java names to property names. As the dot is the most common separator, that one is supported.

> Defining a new property name with a fallback to the old one is only possible, but not so nice. The basic idea is that you define your annotation with property names not using a dash, then in your activate method (or if you're using constructor injection in your constructor) you have at least two parameters, one uses your annotation type, the other one is a map which then gets all configuration properties. Now, you need to check whether the old property is set in the map and use that value - but only if the new property is not set in the map. Otherwise, you can use the value from the annotation.

**Comment by badvision** [comment]
Makes sense for maximum compatibility but this also adds a significant amount of additional technical debt.

**Comment by davidjgonzalez** [comment]
lets try to design a solution to minimized additional technical debt then..   we made this bed, so..

**Comment by badvision** [comment]
I'm all ears, dude. 😉

**Comment by royteeuwen** [comment]
why do you say this does not work? Are you sure that you can't use dashes? did you try doing my$_$property() ? See following documentation:

https://osgi.org/specification/osgi.cmpn/7.0.0/service.component.html#service.component-component.property.types

**Comment by davidjgonzalez** [comment]
@royteeuwen see the discussion at: https://github.com/Adobe-Consulting-Services/acs-aem-commons/issues/1567

R7 is only available on 6.4 and ACS Commons supports the current and prior viersion of AEM - so at the time of writing, AEM 6.4 and 6.3.

**Comment by heervisscher** [comment]
Also wondering why there is not an 'internalName' attribute in @ AttributeDefinition, to me this make it also more clear while reading the code what the prop-name is.

**Comment by joerghoh** [comment]
in #1567  I evaluated 2 options how we could deal with it. 
* Switching to R7 annotations does not help, because the annotations write a metatype 1.4 requirement into the XML, which AEM 6.3 cannot handle (6.4 can).
* Sticking with a mixed approach (stay with the old SCR annotations where we would otherwise introduce changes into the property naming) breaks the jacoco/coveralls plugins.

From my point of view this leaves us with 2 options:

1. Reverting the switch to R6 annotations and go with the SCR annotations for the next year, until we drop support of AEM 6.3
2. We analyze and fix the jacoco stuff to get the test coverage report back.

While (1) seems the easiest way, we would throw away quite some work.  Or we find someone who wants to anaylze jacoco.

WDYT?

**Comment by justinedelson** [comment]
@joerghoh this might a PITA, but did you try specifying the `xmlns` attribute on the `@Component` annotation?

Actually, a global find-and-replace for `@Component(` replacing it with `@Component(xmlns=<1.3 NS URL>,` could work.

**Comment by justinedelson** [comment]
maybe just try one to see if it actually generates the right xml file

**Comment by heervisscher** [comment]
Here my thoughts:

1) we leave the code as is, so no more '-' in the property names.
2) we can do a check in the activate and check for '-' in the property-names, then raise something in the log

I really don't like method signatures with $_$, also it will be very hard for people to know the property-names based on the code.

**Comment by badvision** [comment]
Ok so some time has passed and there's no better alternative suggested.  Sounds like the path of least resistance is to go with R7 annotations, at least for the services which use hyphens in the property names.

**Comment by kwin** [comment]
Sorry for joining the party late, but I guess there is one misunderstanding here: Although the `$_$` support has only been added with the OSGi R7 specification this only refers to tools generating the metadata (`bnd` in our case). Using this pattern does not(!) lead to a runtime dependency towards OSGi R7, at runtime everything is still compatible with DS 1.3 (i.e. OSGi R6!). `bnd` since version 4.0 (and by that maven-bundle-plugin 4.0+) supports that new mapping!

**Comment by justinedelson** [comment]
@kwin I don't think this is correct.  What I've seen is that once you use use $_$ in a property method, the generated descriptor uses the namespace schema for 1.4 and thus is not compatible with 1.3. Are you doing something different?

**Comment by kwin** [comment]
AFAIK the DS namespace should now primarily depend on the dependency towards the annotation (see the discussion at https://github.com/bndtools/bnd/issues/2163). But if it really always generates the NS 1.4 schema, this seems like a bug to me and should be reported at bnd.

**Comment by justinedelson** [comment]
This is clearly by design. See https://github.com/bndtools/bnd/blob/master/biz.aQute.bndlib/src/aQute/bnd/metatype/OCDReader.java#L540. Further, I believe *has* to require 1.4 at runtime since otherwise the runtime doesn't understand how to map the property (for example) `conflict-resolution.default` to `conflict$_$resolution_default()`. The only way that you could possibly make this work is to *only* use the annotation to declare the properties, but actually use the `Map<String, Object>` signature on the activate method. But again, that would require either changing the behavior of bnd (which seems unlikely to me) or post-processing the XML files. I tried the latter approach here: https://github.com/justinedelson/acs-aem-commons/commit/34b65aad6038ba77a571b9c2578a60fa5b1510df and it didn't work at runtime.

**Comment by kwin** [comment]
@justinedelson You are right, indeed this requires DS 1.4 at runtime.

**Comment by adamcin** [comment]
What if we kept the no-hyphen property names wrt the generated metatype, but also switched the hyphen-affected services back to the Map<> or ComponentContext signature activate methods to get access to the existing hyphenated config properties at runtime to treat them as deprecated defaults for the new non-hyphenated versions of the properties?

**Comment by davidjgonzalez** [comment]
@adamcin There's some GH issue (so many now!) where we talked though ~this concept. 

 TL;DR it increases code complexity, requires communication/documentation and asks users to update their OSGi configs (at some point). With 6.5 right around the corner, we'll be back in our self-imposed current and previous GA version support model. 

Don't get me wrong, I don't _like_ any of these solutions, but I 100% believe this is the right big-picture choice for the project. We can always cherrypick important fixes back to a 3.x line (not pleasant, but doable)

If you're interested here is the working list of items to get 4.0.0 out the door: https://github.com/Adobe-Consulting-Services/acs-aem-commons/issues/1629

**Comment by adamcin** [comment]
Okay, so we'll move forward with `$_$`.

**Comment by davidjgonzalez** [comment]
For posterity, `$_$` does not work in AEM 6.4 due to incorrect installed metatype bundle.

That said, closing due to: https://github.com/Adobe-Consulting-Services/acs-aem-commons/issues/1741
